AppDB/backup/backup_recovery_helper.py

```diff
@@ -211,3 +227,129 @@ def untar_backup_files(source):
     raise backup_exceptions.BRException(
       "Exception while untarring backup file '{0}'.".format(source))
   logging.info("Done untarring '{0}'.".format(source))
+
+def app_backup(storage, full_bucket_name=None):
+  """ Saves the app source code at the backups location on the filesystem.
+
+  Args:
+    storage: A str, one of the StorageTypes class members.
+    full_bucket_name: A str, the name of the backup file to upload to remote
+      storage.
+  Returns:
+    True on success, False otherwise.
+  """
+  for dir_path, _, filenames in os.walk(APP_DIR_LOCATION):
+    for filename in filenames:
+      # Copy source code tars to backups location.
+      source = '{0}/{1}'.format(dir_path, filename)
+      destination = '{0}/{1}'.format(APP_BACKUP_DIR_LOCATION, filename)
+      try:
+        shutil.copy(source, destination)
+      except:
+        logging.error("Error while backing up '{0}'. ".format(source))
+        delete_app_tars(APP_BACKUP_DIR_LOCATION)
+        return False
+
+      # Upload to GCS.
+      if storage == StorageTypes.GCS:
+        source = '{0}/{1}'.format(APP_DIR_LOCATION, filename)
+        destination = '{0}/apps/{1}'.format(full_bucket_name, filename)
+        logging.debug("Destination: {0}".format(destination))
+        if not gcs_helper.upload_to_bucket(destination, source):
+          logging.error("Error while uploading '{0}' to GCS. ".format(source))
+          delete_app_tars(APP_BACKUP_DIR_LOCATION)
+          return False
+  return True
+
+def app_restore(storage, bucket_name=None):
+  """ Restores the app source code from the backups location on the filesystem.
+
+  Args:
+    storage: A str, one of the StorageTypes class members.
+    bucket_name: A str, the name of the bucket to restore apps from.
+  Returns:
+    True on success, False otherwise.
+  """
+  # Download from GCS to backups location.
+  if storage == StorageTypes.GCS:
+    objects = gcs_helper.list_bucket(bucket_name)
+    for app_path in objects:
+      if not app_path.startswith('apps/'):
+        continue
+
+      # Only keep the relative name of the app file.
+      # E.g. myapp.tar.gz (app_file) out of apps/myapp.tar.gz (app_path)
+      app_file = app_path[5:]
```

Comment:
I would suggest making 'apps/' a constant and getting the length of it for this slicing.